api/v0/ingest/schema/envelope.go

```diff
@@ -0,0 +1,101 @@
+package ingestion
+
+import (
+	"bytes"
+	"errors"
+
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	crypto "github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/record"
+	mh "github.com/multiformats/go-multihash"
+)
+
+const (
+	adSignatureCodec  = "/indexer/ingest/adSginature"
+	adSignatureDomain = "indexer"
+)
+
+type advSignatureRecord struct {
+	domain *string
+	codec  []byte
+	advID  []byte
+}
+
+func (r *advSignatureRecord) Domain() string {
+	if r.domain != nil {
+		return *r.domain
+	}
+	return adSignatureDomain
+}
+
+func (r *advSignatureRecord) Codec() []byte {
+	if r.codec != nil {
+		return r.codec
+	}
+	return []byte(adSignatureCodec)
+}
+
+func (r *advSignatureRecord) MarshalRecord() ([]byte, error) {
+	return r.advID, nil
+}
+
+func (r *advSignatureRecord) UnmarshalRecord(buf []byte) error {
+	r.advID = buf
+	return nil
+}
+
+// Generates advertisements ID from the data included.
+func genAdvertisementID(indexID Link_Index, provider string, previousAdvID []byte) ([]byte, error) {
+	lindex, err := indexID.AsLink()
+	if err != nil {
+		return nil, err
+	}
+	bindex := lindex.(cidlink.Link).Cid.Bytes()
+	// Signature data is indexID+provider+previousAdvID
+	sigData := make([]byte, len(bindex)+len([]byte(provider))+len(previousAdvID))
+	copy(sigData[:len(bindex)], bindex)
+	copy(sigData[len(bindex):], []byte(provider))
+	copy(sigData[len(bindex)+len(provider):], previousAdvID)
+
+	return mh.Encode(sigData, mhCode)
```

Comment:
consider using a codec / ipld encoder over a struct rather than doing this manually?

Reply:
Can you elaborate on this a bit? I don't think there is any encoding to do here, I am just concatenating the three fields in bytes to compute their multihash in order to use the result as the ID for the advertisement. Thanks!

Reply:
you have a serialization format of how to go from a structured object to a byte-stream, and then get a cid of it. That's what a codec does in getting a link out of an ipld object, but in a general way, i think

Reply:
Ok, gotcha! Thanks. 

As AdvertisementID != AdvertisementLink because the the latter includes the signature and AdvertisementID only needs three of the fields of the node and is used to create the signature, I thought it made sense to do it this way, but I'll check if it is better to have an IPLD node with the three fields, and use its encoded link to create the signature.